abook.1
.TH ABOOK 1 2006-09-06
.nh
.SH NAME
abook \- text-based address book program
.SH SYNOPSIS
.B abook
[ \fIOPTION\fR ]
.SH DESCRIPTION
This manual page documents briefly the
.B abook
program.
This manual page was written for the Debian GNU/Linux distribution
because the original program does not have a manual page.
.PP
.B abook 
is a text-based address book program. It contains Name, Email, Address 
and various Phone fields. It is designed for use with mutt, but can be
equally useful on its own.
.SH OPTIONS
.TP
\fB\-h \-\-help\fP
Show usage.
.TP
\fB\-C \-\-config\fP \fI<filename>\fR
Use an alternative configuration file (default is \fI$HOME/.abook/abookrc\fR).
.TP
\fB\-\-datafile\fP \fI<filename>\fR
Use an alternative addressbook file (default is \fI$HOME/.abook/addressbook\fR).
.TP
\fB\-\-mutt\-query\fP \fI<string>\fR [ \fB\-\-outformat\fP \fI<outputformat>\fR ]
Make a query for mutt (search the addressbook for \fI<string>\fR).
.br
The \fB\-\-datafile\fP option, as documented above, may be used
.BI BEFORE
this option to search a different addressbook file.
.br
Only a subset of the below \fI<outputformat>\fR are allowed: \fBmutt\fP (default), \fBvcard\fP and \fBcustom\fP
.TP
\fB\-\-convert\fP [ \fB\-\-informat\fP \fI<inputformat>\fR ] [ \fB\-\-infile\fP \fI<inputfile>\fR ] [ \fB\-\-outformat\fP \fI<outputformat>\fR ] [ \fB\-\-outfile\fP \fI<outputfile>\fR ]
Converts \fI<inputfile>\fR in \fI<inputformat>\fR to \fI<outputfile>\fR in \fI<outputformat>\fR
(defaults are \fBabook\fP, \fBstdin\fP, \fBtext\fP and \fBstdout\fP).

.br
The following \fIinputformats\fR are supported:
.br
- \fBabook\fP abook native format
.br
- \fBldif\fP ldif / Netscape addressbook
.br
- \fBmutt\fP mutt alias
.br
- \fBpine\fP pine addressbook
.br
- \fBcsv\fP comma separated values
.br
- \fBpalmcsv\fP Palm comma separated values
.br
- \fBvcard\fP VCard addressbook

.br
The following \fIoutputformats\fR are supported:
.br
- \fBabook\fP abook native format
.br
- \fBldif\fP ldif / Netscape addressbook (.4ld)
.br
- \fBmutt\fP mutt alias
.br
- \fBhtml\fP html document
.br
- \fBpine\fP pine addressbook
.br
- \fBvcard\fP VCard addressbook
.br
- \fBcsv\fP comma separated values
.br
- \fBpalmcsv\fP Palm comma separated values
.br
- \fBelm\fP elm alias
.br
- \fBtext\fP plain text
.br
- \fBspruce\fP Spruce address book
.br
- \fBwl\fP Wanderlust address book
.br
- \fBbsdcal\fP BSD calendar
.br
- \fBcustom\fP Custom output format, see below
.TP
\fB\-\-outformatstr\fP \fI<string>\fR
Only used if \fB\-\-mutt\-query\fP \fIor\fR \fB\-\-convert\fP is specified \fIand\fR \fB\-\-outformat\fP=\fIcustom\fR. \fI<string>\fR is a format string allowing placeholders.
.br
A placeholder can be any of the standard fields names (see \fBabookrc\fP(5)) and must be encapsulated by curly brackets.
.br
The default value is "{nick} ({name}): {mobile}"
.br
If \fI<string>\fR starts with \fI!\fR only entries whose all fields from \fI<string>\fR are non-NULL are included.
.TP
\fB\-\-add-email\fP
Read an e-mail message from stdin and add the sender to the addressbook.
.TP
\fB\-\-add-email-quiet\fP
Same as --add-email but doesn't confirm adding.
.TP
\fB\-\-formats\fP
List available formats.

.SH COMMANDS DURING USE
Press '\fB?\fP' during use to get a list of commands.
.SH SEE ALSO
.BR mutt (1),
.BR abookrc (5)
.br
.SH AUTHOR
This manual page was written by Alan Ford <alan@whirlnet.co.uk>,
for the Debian GNU/Linux system (but may be used by others).

.br
.B abook
was written by Jaakko Heinonen <jheinonen@users.sourceforge.net>


ChangeLog
git
 - custom output format (Raphaël Droz)
 - remove duplicates (Fabio Zanini)
 - merge entries (Fabio Zanini)
 - mutt groups support (Fabio Zanini)
 - vcard initial input filter (Michael Krolikowski)
 - vcard parser improvements (Fabio Zanini, Raphaël Droz)
 - color support (Thorsten Wißmann)
 - mouse support (Thorsten Wißmann)
 - extra-fields deletion bugfix (Jorrit Tijben)
 - additinal keybinding (Hagen Fuchs)
 - autotools update (Fabio Zanini)

0.6.0
 - configurable views (Cedric Duval)
 - new field types, lists and improved custom field support (Cedric Duval)
 - index_format option (Cedric Duval)
 - Italian translation (Claudio Stazzone)

0.5.6
 - translation fixes
 - fixed gcrd export format (\r\n line endings according to RFC, Marc Tardif)
 - fix localized keybindings with --disable-nls

0.5.5
 - fixed gcrd export format (Marc Tardif)
 - i18n support (Cedric Duval)
 - French translation (Cedric Duval)
 - Swedish translation (Susanna Björverud)
 - Japanese translation (Takahashi Tamotsu)
 - German translation (Johannes Weißl)
 - palmcsv import filter (Marc Tardif)
 - use better common code for csv filters (Marc Tardif)
 - translation HOWTO (Cedric Duval)
 - wanderlust export filter (Josef Schugt)
 - create a backup at save time
 - editor undo fix
 - ui_readline limit fix

0.5.4
 - add show_cursor config option (idea from Cheryl Homiak)
 - autoconf/automake update
 - replace abook_malloc, abook_realloc and my_free with new xmalloc routines
 - replace strdup with xstrdup
 - update for abook_rl (abook_readline)
 - make it compile with tcc (Tiny C Compiler)
 - code cleanups and minor fixes

0.5.3
 - add allcvs filters (Christoph Sobotka)
 - removed a lefover debug message
 - fixed mutt import filter nickname cutoff
 - wrap searches
 - contrib updates

0.5.2
 - five custom fields added
 - experimental wide character support
 - fixed a crash bug in mutt filter
 - fixed errors in configure.in (especially snprintf-functions were incorrectly
   detected)
 - abookrc manual page update

0.5.1
 - sort by field command
 - duplicate item command
 - quit without saving command
 - autoconf/automake update
 - minor filter fixes
 - tried to fix \" quotation problem with mutt import filter
 - removed obsolete filesel_sort option from abookrc manual page
 - added bbdb2xx translator source to contrib

0.5.0
 - tried to work around some problems with readline
 - some documentation update
20030530
 - updated autoconf/automake
 - updated man page
 - pine import filter update
20021008
 - options.c rewritten (new mutt style rc file support)
 - palmcsv export filter (Koenraad Heijlen)
 - added a spec file for rpm support
 - added --with-readline configure option
 - merged newtab patch by Sinan Kaan Yerli
 - new option: add_email_prevent_duplicates
 - bugfixes
2002-03-11
 - readline support
 - abook can be compiled with g++
 - attempt to improve --datafile behavior
 - html filter update (Morten Brix Pedersen)
 - store rcfile and addressbook to .abook directory
2002-02-04
 - use getopt to parse command line (also new options added)
 - use strcoll instead of strcmp for sorting entries
 - html filter fix
 - bugfixes

0.4.17
 - fixed an annoying file creation bug

0.4.16
 - use strcoll instead of strcmp for sorting entries
 - html filter fix
 - many bugfixes

0.4.15
 - new options --add-email and --add-email-quiet
 - proper mutt alias import filter
 - editor update
 - don't handle extra_column and extra_alternative as numbers in abookrc
 - added second address field
 - mail.vim update in contrib directory

0.4.14
 - added csv import filter
 - csv export filter fix
 - minor html filter update
 - fixed a minor memory leak in pine import filter
 - filter cleanups
 - fixes
 - "quit and print selected item(s) to stderr" command (Brian Medley)
 - contrib update

0.4.13
 - major code reorganization (phase 1)
 - fixed help viewer functions
 - ability to disable use of non-ascii characters (configuration option)
 - changed SIGINT behaviour
 - html export filter creates now valid html documents
 - ability to print/export only selected items
 - new editor commands
 - text filter update
 - query/find code cleanup
 - fixed snprintf related bugs
 - minor bugfixes

0.4.12
 - added man page for abookrc (Alan Ford)
 - notes are returned as optional data in mutt queries
 - created a workaround for a mutt query bug
 - new "extra_alternative" config option (Alan)
 - added undo feature to editor
 - FAQ
 - macro updates
 - fixed the cancel key behavior in editor

0.4.11
 - new configuration file format
 - new sorting routines
 - minor mutt query patch (Matt)
 - html filter update
 - added mail2abook.py script to improve the mutt integration
 - memory allocation changes
 - contrib update

0.4.10
 - mutt alias import filter
 - --datafile option should work better now
 - merged HPUX patch (submitted by Matt Kraai)
 - added field for nickname/alias
 - bugfixes

0.4.9
 - increased the maximum length of an e-mail address
 - added Spruce and ldif export filters
 - text filter update
 - added support for systemwide configuration file (/etc/abook.conf)
 - list.c cleanup
 - minor fixes

0.4.8
 - new enter_string function (scrollable input)
 - fileselector
 - new fields for phone numbers
 - extra field support
 - added open database command and --datafile option

0.4.7
 - fixed terminal resizing bug
 - new man page
 - fixed "roll emails" bug
 - merged Alan's urlview patch
 - fixed some compile warnings on some systems

0.4.6
 - new proper ldif import filter
 - printing
 - GnomeCard export filter
 - improved ui
 - csv export filter
 - elm alias export filter
 - plain text export filter
 - improved import / export system
 - a man page (contributed by Alan Ford)
 - some fixes

0.4.5
 - new command line option --convert to make format conversions
 - new common functions for filters
 - mutt alias export filter
 - minor filter updates
 - fixed a compile problem in list.c 

0.4.4
 - fixed a serious segfault bug in editor
 - conff.c update

0.4.3

abook.c
/*
 * $Id$
 *
 * by JH <jheinonen@users.sourceforge.net>
 *
 * Copyright (C) Jaakko Heinonen
 */

#include <errno.h>
#include <fcntl.h>
#include <ctype.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#ifdef HAVE_CONFIG_H
#	include "config.h"
#endif
#if defined(HAVE_LOCALE_H) && defined(HAVE_SETLOCALE)
#	include <locale.h>
#endif
#include <assert.h>
#include "abook.h"
#include "gettext.h"
#include "ui.h"
#include "database.h"
#include "list.h"
#include "filter.h"
#include "edit.h"
#include "misc.h"
#include "options.h"
#include "getname.h"
#include "getopt.h"
#include "views.h"
#include "xmalloc.h"

static void             init_abook();
static void		quit_abook_sig(int i);
static void             set_filenames();
static void             parse_command_line(int argc, char **argv);
static void             show_usage();
static void             mutt_query(char *str);
static void             init_mutt_query();
static void		convert(char *srcformat, char *srcfile,
				char *dstformat, char *dstfile);
static void		add_email(int);

char *datafile = NULL;
static char *rcfile = NULL;

// custom formatting
char custom_format[FORMAT_STRING_LEN] = "{nick} ({name}): {mobile}";
char *parsed_custom_format = NULL;
enum field_types *custom_format_fields = 0;
struct abook_output_item_filter selected_item_filter;

bool alternative_datafile = FALSE;
bool alternative_rcfile = FALSE;


static int
datafile_writeable()
{
	FILE *f;

	assert(datafile != NULL);

	if( (f = fopen(datafile, "a")) == NULL)
		return FALSE;

	fclose(f);

	return TRUE;
}

static void
check_abook_directory()
{
	struct stat s;
	char *dir;

	assert(!is_ui_initialized());

	if(alternative_datafile)
		return;

	dir = strconcat(getenv("HOME"), "/" DIR_IN_HOME, NULL);
	assert(dir != NULL);

	if(stat(dir, &s) == -1) {
		if(errno != ENOENT) {
			perror(dir);
                        free(dir);
                        exit(EXIT_FAILURE);
		}
		if(mkdir(dir, 0700) == -1) {
			printf(_("Cannot create directory %s\n"), dir);
			perror(dir);
			free(dir);
			exit(EXIT_FAILURE);
		}
	} else if(!S_ISDIR(s.st_mode)) {
		printf(_("%s is not a directory\n"), dir);
		free(dir);
		exit(EXIT_FAILURE);
	}

	free(dir);
}

static void
xmalloc_error_handler(int err)
{
	/*
	 * We don't try to save addressbook here because we don't know
	 * if it's fully loaded to memory.
	 */
	if(is_ui_initialized())
		close_ui();

	fprintf(stderr, _("Memory allocation failure: %s\n"), strerror(err));
	exit(EXIT_FAILURE);
}

static void
init_abook()
{
	set_filenames();
	check_abook_directory();
	init_opts();
	if(load_opts(rcfile) > 0) {
		printf(_("Press enter to continue...\n"));
		fgetc(stdin);
	}
	init_default_views();

	signal(SIGTERM, quit_abook_sig);

	init_index();

	if(init_ui())
		exit(EXIT_FAILURE);

	umask(DEFAULT_UMASK);

	if(!datafile_writeable()) {
		char *s = strdup_printf(_("File %s is not writeable"), datafile);
		refresh_screen();
		statusline_msg(s);
		free(s);
		if(load_database(datafile) || !statusline_ask_boolean(
					_("If you continue all changes will "
				"be lost. Do you want to continue?"), FALSE)) {
			free_opts();
			/*close_database();*/
			close_ui();
			exit(EXIT_FAILURE);
		}
	} else
		load_database(datafile);

	refresh_screen();
}

void
quit_abook(int save_db)
{
	if(save_db)  {
		if(opt_get_bool(BOOL_AUTOSAVE))
			save_database();
		else if(statusline_ask_boolean(_("Save database"), TRUE))
			save_database();
	} else if(!statusline_ask_boolean(_("Quit without saving"), FALSE))
		return;

	free_opts();
	close_database();

	close_ui();

	exit(EXIT_SUCCESS);
}

static void
quit_abook_sig(int i)
{
	quit_abook(QUIT_SAVE);
}

int
main(int argc, char **argv)
{
#if defined(HAVE_SETLOCALE) && defined(HAVE_LOCALE_H)
	setlocale(LC_MESSAGES, "");
	setlocale(LC_TIME, "");
	setlocale(LC_CTYPE, "");
	setlocale(LC_COLLATE, "");
#endif
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	xmalloc_set_error_handler(xmalloc_error_handler);

	prepare_database_internals();

	parse_command_line(argc, argv);

	init_abook();

	get_commands();

	quit_abook(QUIT_SAVE);

	return 0;
}

static void
free_filenames()
{
	xfree(rcfile);
	xfree(datafile);
}


static void
set_filenames()
{
	struct stat s;

	if( (stat(getenv("HOME"), &s)) == -1 || ! S_ISDIR(s.st_mode) ) {
		fprintf(stderr,_("%s is not a valid HOME directory\n"), getenv("HOME") );
		exit(EXIT_FAILURE);
	}

	if(!datafile)
		datafile = strconcat(getenv("HOME"), "/" DIR_IN_HOME "/"
				DATAFILE, NULL);

	if(!rcfile)
		rcfile = strconcat(getenv("HOME"), "/" DIR_IN_HOME "/"
				RCFILE, NULL);

	atexit(free_filenames);
}

/*
 * CLI
 */

enum {
	MODE_CONT,
	MODE_ADD_EMAIL,
	MODE_ADD_EMAIL_QUIET,
	MODE_QUERY,
	MODE_CONVERT
};

static void
change_mode(int *current, int mode)
{
	if(*current != MODE_CONT) {
		fprintf(stderr, _("Cannot combine options --mutt-query, "
				"--convert, "
				"--add-email or --add-email-quiet\n"));
		exit(EXIT_FAILURE);
	}

	*current = mode;
}

void
set_filename(char **var, char *path)
{
	char *cwd;

	assert(var != NULL);
	assert(*var == NULL); /* or else we probably leak memory */
	assert(path != NULL);

	if(*path == '/') {
		*var = xstrdup(path);
		return;
	}

	cwd = my_getcwd();

	*var = strconcat(cwd, "/", path, NULL);

	free(cwd);
}

#define set_convert_var(X) do { if(mode != MODE_CONVERT) {\
	fprintf(stderr, _("please use option --%s after --convert option\n"),\
			long_options[option_index].name);\
		exit(EXIT_FAILURE);\
	} else\
		X = optarg;\
	} while(0)

static void
parse_command_line(int argc, char **argv)
{
	int mode = MODE_CONT;
	char *query_string = NULL;
	char *informat = "abook",
		*outformat = "text",
		*infile = "-",
		*outfile = "-";
	int c;
	selected_item_filter = select_output_item_filter("muttq");

	for(;;) {
		int option_index = 0;
		enum {
			OPT_ADD_EMAIL,
			OPT_ADD_EMAIL_QUIET,
			OPT_MUTT_QUERY,
			OPT_CONVERT,
			OPT_INFORMAT,
			OPT_OUTFORMAT,
			OPT_OUTFORMAT_STR,
			OPT_INFILE,
			OPT_OUTFILE,
			OPT_FORMATS
		};
		static struct option long_options[] = {
			{ "help", 0, 0, 'h' },
			{ "add-email", 0, 0, OPT_ADD_EMAIL },
			{ "add-email-quiet", 0, 0, OPT_ADD_EMAIL_QUIET },
			{ "datafile", 1, 0, 'f' },
			{ "mutt-query", 1, 0, OPT_MUTT_QUERY },
			{ "config", 1, 0, 'C' },
			{ "convert", 0, 0, OPT_CONVERT },
			{ "informat", 1, 0, OPT_INFORMAT },
			{ "outformat", 1, 0, OPT_OUTFORMAT },
			{ "outformatstr", 1, 0, OPT_OUTFORMAT_STR },
			{ "infile", 1, 0, OPT_INFILE },
			{ "outfile", 1, 0, OPT_OUTFILE },
			{ "formats", 0, 0, OPT_FORMATS },
			{ 0, 0, 0, 0 }
		};

		c = getopt_long(argc, argv, "hC:",
				long_options, &option_index);

		if(c == -1)
			break;

		switch(c) {
			case 'h':
				show_usage();
				exit(EXIT_SUCCESS);
			case OPT_ADD_EMAIL:
				change_mode(&mode, MODE_ADD_EMAIL);
				break;
			case OPT_ADD_EMAIL_QUIET:
				change_mode(&mode, MODE_ADD_EMAIL_QUIET);
				break;
			case 'f':
				set_filename(&datafile, optarg);
				alternative_datafile = TRUE;
				break;
			case OPT_MUTT_QUERY:
				query_string = optarg;
				change_mode(&mode, MODE_QUERY);
				break;
			case 'C':
				set_filename(&rcfile, optarg);
				alternative_rcfile = TRUE;
				break;
			case OPT_CONVERT:
				change_mode(&mode, MODE_CONVERT);
				break;
			case OPT_INFORMAT:
				set_convert_var(informat);
				break;
			case OPT_OUTFORMAT:
				// ascii-name is stored, it's used to traverse
				// e_filters[] in MODE_CONVERT (see export_file())
				outformat = optarg;
				// but in case a query-compatible filter is requested
				// try to guess right now which one it is, from u_filters[]
				selected_item_filter = select_output_item_filter(outformat);
				break;
			case OPT_OUTFORMAT_STR:
				strncpy(custom_format, optarg, FORMAT_STRING_LEN - 1);
				custom_format[FORMAT_STRING_LEN] = 0;
				break;
			case OPT_INFILE:
				set_convert_var(infile);
				break;
			case OPT_OUTFILE:
				set_convert_var(outfile);
				break;
			case OPT_FORMATS:
				print_filters();
				exit(EXIT_SUCCESS);
			default:
				exit(EXIT_FAILURE);
		}
	}

	// if the output format requested does not allow filtered querying
	// (not in u_filter[]) and --convert has not been specified; bailout
	if(! selected_item_filter.func && mode != MODE_CONVERT) {
	  printf("output format %s not supported or incompatible with --mutt-query\n", outformat);
	  exit(EXIT_FAILURE);
	}
	if(! selected_item_filter.func)
		selected_item_filter = select_output_item_filter("muttq");
	else if (! strcmp(outformat, "custom") && *custom_format) {
		parsed_custom_format = (char *)malloc(FORMAT_STRING_LEN * sizeof(char*));
		custom_format_fields = (enum field_types *)malloc(FORMAT_STRING_MAX_FIELDS * sizeof(enum field_types *));
		parse_custom_format(custom_format, parsed_custom_format, custom_format_fields);
	}
	if(optind < argc) {
		fprintf(stderr, _("%s: unrecognized arguments on command line\n"),
				argv[0]);
		exit(EXIT_FAILURE);
	}

	switch(mode) {
		case MODE_ADD_EMAIL:
			add_email(0);
		case MODE_ADD_EMAIL_QUIET:
			add_email(1);
		case MODE_QUERY:
			mutt_query(query_string);
		case MODE_CONVERT:
			convert(informat, infile, outformat, outfile);
	}
}


static void
show_usage()
{
	puts	(PACKAGE " v " VERSION "\n");
	puts	(_("     -h	--help				show usage"));
	puts	(_("     -C	--config	<file>		use an alternative configuration file"));
	puts	(_("	--datafile	<file>		use an alternative addressbook file"));
	puts	(_("	--mutt-query	<string>	make a query for mutt"));
	puts	(_("	--add-email			"
			"read an e-mail message from stdin and\n"
		"					"
		"add the sender to the addressbook"));
	puts	(_("	--add-email-quiet		"
		"same as --add-email but doesn't\n"
		"					require to confirm adding"));
	putchar('\n');
	puts	(_("	--convert			convert address book files"));
	puts	(_("	options to use with --convert:"));
	puts	(_("	--informat	<format>	format for input file"));
	puts	(_("					(default: abook)"));
	puts	(_("	--infile	<file>		source file"));
	puts	(_("					(default: stdin)"));
	puts	(_("	--outformat	<format>	format for output file"));
	puts	(_("					(default: text)"));
	puts	(_("	--outfile	<file>		destination file"));
	puts	(_("					(default: stdout)"));
	puts	(_("	--formats			list available formats"));
}

/*
 * end of CLI
 */


static void
quit_mutt_query(int status)
{
	close_database();
	free_opts();

	exit(status);
}

static void
mutt_query(char *str)
{
	init_mutt_query();

	if( str == NULL || !strcasecmp(str, "all") ) {
		export_file("muttq", "-");
	} else {
		int search_fields[] = {NAME, EMAIL, NICK, -1};
		int i;
		if( (i = find_item(str, 0, search_fields)) < 0 ) {
			printf("Not found\n");
			quit_mutt_query(EXIT_FAILURE);
		}
		putchar('\n');
		while(i >= 0) {
			e_write_item(stdout, i, selected_item_filter.func);
			i = find_item(str, i + 1, search_fields);
		}
	}

	quit_mutt_query(EXIT_SUCCESS);
}

static void
init_mutt_query()
{
	set_filenames();
	init_opts();
	load_opts(rcfile);

	if( load_database(datafile) ) {
		printf(_("Cannot open database\n"));
		quit_mutt_query(EXIT_FAILURE);
		exit(EXIT_FAILURE);
	}
}


static char *
make_mailstr(int item)
{
	char email[MAX_EMAIL_LEN];
	char *ret;
	char *name = strdup_printf("\"%s\"", db_name_get(item));

	get_first_email(email, item);

	ret = *email ?
		strdup_printf("%s <%s>", name, email) :
		xstrdup(name);

	free(name);

	return ret;
}

void
print_stderr(int item)
{
	fprintf (stderr, "%c", '\n');

	if( is_valid_item(item) )
		muttq_print_item(stderr, item);
	else {
		struct db_enumerator e = init_db_enumerator(ENUM_SELECTED);
		db_enumerate_items(e) {
			muttq_print_item(stderr, e.item);
		}
	}

}

void
launch_mutt(int item)
{
	char *cmd = NULL, *mailstr = NULL;
	char *mutt_command = opt_get_str(STR_MUTT_COMMAND);

	if(mutt_command == NULL || !*mutt_command)
		return;

	if( is_valid_item(item) )
		mailstr = make_mailstr(item);
	else {
		struct db_enumerator e = init_db_enumerator(ENUM_SELECTED);
		char *tmp = NULL;
		db_enumerate_items(e) {
			tmp = mailstr;
			mailstr = tmp ?
				strconcat(tmp, ",", make_mailstr(e.item), NULL):
				strconcat(make_mailstr(e.item), NULL);
			free(tmp);
		}
	}

	cmd = strconcat(mutt_command, " \'", mailstr, "\'", NULL);
	free(mailstr);
#ifdef DEBUG
	fprintf(stderr, "cmd: %s\n", cmd);
#endif
	system(cmd);
	free(cmd);

	/*
	 * we need to make sure that curses settings are correct
	 */
	ui_init_curses();
}

void
launch_wwwbrowser(int item)
{
	char *cmd = NULL;

	if( !is_valid_item(item) )
		return;

	if(db_fget(item, URL))
		cmd = strdup_printf("%s '%s'",
				opt_get_str(STR_WWW_COMMAND),
				safe_str(db_fget(item, URL)));
	else
		return;

	if ( cmd )
		system(cmd);

	free(cmd);

	/*
	 * we need to make sure that curses settings are correct
	 */
	ui_init_curses();
}

FILE *
abook_fopen (const char *path, const char *mode)
{
	struct stat s;
	bool stat_ok;

	stat_ok = (stat(path, &s) != -1);

	if(strchr(mode, 'r'))
		return (stat_ok && S_ISREG(s.st_mode)) ?
			fopen(path, mode) : NULL;
	else
		return (stat_ok && S_ISDIR(s.st_mode)) ?
			NULL : fopen(path, mode);
}

static void
convert(char *srcformat, char *srcfile, char *dstformat, char *dstfile)
{
	int ret=0;

	if( !srcformat || !srcfile || !dstformat || !dstfile ) {
		fprintf(stderr, _("too few arguments to make conversion\n"));
		fprintf(stderr, _("try --help\n"));
	}

#ifndef DEBUG
	if( !strcasecmp(srcformat, dstformat) ) {
		printf(	_("input and output formats are the same\n"
			"exiting...\n"));
		exit(EXIT_FAILURE);
	}
#endif

	set_filenames();
	init_opts();
	load_opts(rcfile);
	init_standard_fields();

	switch(import_file(srcformat, srcfile)) {
		case -1:
			fprintf(stderr,
				_("input format %s not supported\n"), srcformat);
			ret = 1;
			break;
		case 1:
			fprintf(stderr, _("cannot read file %s\n"), srcfile);
			ret = 1;
			break;
	}

	if(!ret)
		switch(export_file(dstformat, dstfile)) {
			case -1:
				fprintf(stderr,
					_("output format %s not supported\n"),
					dstformat);
				ret = 1;
				break;
			case 1:
				fprintf(stderr,
					_("cannot write file %s\n"), dstfile);
				ret = 1;
				break;
		}

	close_database();
	free_opts();
	exit(ret);
}

/*
 * --add-email handling
 */

static int add_email_count = 0;

static void
quit_add_email()
{
	if(add_email_count > 0) {
		if(save_database() < 0) {
			fprintf(stderr, _("cannot open %s\n"), datafile);
			exit(EXIT_FAILURE);
		}
		printf(_("%d item(s) added to %s\n"), add_email_count, datafile);
	} else {
		puts(_("Valid sender address not found"));
	}

	exit(EXIT_SUCCESS);
}

static void
quit_add_email_sig(int signal)
{
	quit_add_email();
}

static void
init_add_email()
{
	set_filenames();
	check_abook_directory();
	init_opts();
	load_opts(rcfile);
	init_standard_fields();
	atexit(free_opts);

	/*
	 * we don't actually care if loading fails or not
	 */
	load_database(datafile);

	atexit(close_database);

	signal(SIGINT, quit_add_email_sig);
}

static int
add_email_add_item(int quiet, char *name, char *email)
{
	list_item item;

	if(opt_get_bool(BOOL_ADD_EMAIL_PREVENT_DUPLICATES)) {
		int search_fields[] = { EMAIL, -1 };
		if(find_item(email, 0, search_fields) >= 0) {
			if(!quiet)
				printf(_("Address %s already in addressbook\n"),
						email);
			return 0;
		}
	}

	if(!quiet) {
		FILE *in = fopen("/dev/tty", "r");
		char c;
		if(!in) {
			fprintf(stderr, _("cannot open /dev/tty\n"
				"you may want to use --add-email-quiet\n"));
			exit(EXIT_FAILURE);
		}

		do {
			printf(_("Add \"%s <%s>\" to %s? (%c/%c)\n"),
					name,
					email,
					datafile,
					*S_("keybinding for yes|y"),
					*S_("keybinding for no|n"));
			c = tolower(getc(in));
			if(c == *S_("keybinding for no|n")) {
				fclose(in);
				return 0;
			}
		} while(c != *S_("keybinding for yes|y"));
		fclose(in);
	}

	item = item_create();
	item_fput(item, NAME, xstrdup(name));
	item_fput(item, EMAIL, xstrdup(email));
	add_item2database(item);
	item_free(&item);

	return 1;
}

static void
add_email(int quiet)
{
	char *line;
	char *name = NULL, *email = NULL;
	struct stat s;

	if( (fstat(fileno(stdin), &s)) == -1 || S_ISDIR(s.st_mode) ) {
		fprintf(stderr, _("stdin is a directory or cannot stat stdin\n"));
		exit(EXIT_FAILURE);
	}

	init_add_email();

	do {
		line = getaline(stdin);
		if(line && !strncasecmp("From:", line, 5) ) {
			getname(line, &name, &email);
			add_email_count += add_email_add_item(quiet,
					name, email);
			xfree(name);
			xfree(email);
		}
		xfree(line);
	} while( !feof(stdin) );

	quit_add_email();
}

/*
 * end of --add-email handling
 */
